reactor-core/src/main/java/reactor/util/retry/RetryBuilder.java

```diff
@@ -0,0 +1,375 @@
+	 * @see #andDelayRetryWith(Function) andDelayRetryWith for an asynchronous version
+	 */
+	public RetryBuilder andDoBeforeRetry(
+			Consumer<RetrySignal> doBeforeRetry) {
+		return new RetryBuilder(
+				this.maxAttempts,
+				this.throwablePredicate,
+				this.isTransientErrors,
+				this.doPreRetry.andThen(doBeforeRetry),
+				this.doPostRetry,
+				this.asyncPreRetry,
+				this.asyncPostRetry,
+				this.retryExhaustedGenerator);
+	}
+
+	/**
+	 * Add synchronous behavior to be executed <strong>after</strong> the retry trigger is emitted in
+	 * the companion publisher. This should not be blocking, as the companion publisher
+	 * might be publishing events in a shared thread.
+	 *
+	 * @param doAfterRetry the synchronous hook to execute after retry trigger is started
+	 * @return a new copy of the builder which can either be further configured or used as {@link Retry}
+	 * @see #andRetryThen(Function) andRetryThen for an asynchronous version
+	 */
+	public RetryBuilder andDoAfterRetry(Consumer<RetrySignal> doAfterRetry) {
+		return new RetryBuilder(
+				this.maxAttempts,
+				this.throwablePredicate,
+				this.isTransientErrors,
+				this.doPreRetry,
+				this.doPostRetry.andThen(doAfterRetry),
+				this.asyncPreRetry,
+				this.asyncPostRetry,
+				this.retryExhaustedGenerator);
+	}
+
+	/**
+	 * Add asynchronous behavior to be executed <strong>before</strong> the current retry trigger in the companion publisher,
+	 * delaying the resulting retry trigger with the additional {@link Mono}.
+	 *
+	 * @param doAsyncBeforeRetry the asynchronous hook to execute before original retry trigger is emitted
+	 * @return a new copy of the builder which can either be further configured or used as {@link Retry}
+	 */
+	public RetryBuilder andDelayRetryWith(
+			Function<RetrySignal, Mono<Void>> doAsyncBeforeRetry) {
+		return new RetryBuilder(
+				this.maxAttempts,
+				this.throwablePredicate,
+				this.isTransientErrors,
+				this.doPreRetry,
+				this.doPostRetry,
+				(rs, m) -> asyncPreRetry.apply(rs, m).then(doAsyncBeforeRetry.apply(rs)),
+				this.asyncPostRetry,
+				this.retryExhaustedGenerator);
+	}
+
+	/**
+	 * Add asynchronous behavior to be executed <strong>after</strong> the current retry trigger in the companion publisher,
+	 * delaying the resulting retry trigger with the additional {@link Mono}.
+	 *
+	 * @param doAsyncAfterRetry the asynchronous hook to execute after original retry trigger is emitted
+	 * @return a new copy of the builder which can either be further configured or used as {@link Retry}
+	 */
+	public RetryBuilder andRetryThen(
+			Function<RetrySignal, Mono<Void>> doAsyncAfterRetry) {
+		return new RetryBuilder(
+				this.maxAttempts,
+				this.throwablePredicate,
+				this.isTransientErrors,
+				this.doPreRetry,
+				this.doPostRetry,
+				this.asyncPreRetry,
+				(rs, m) -> asyncPostRetry.apply(rs, m).then(doAsyncAfterRetry.apply(rs)),
+				this.retryExhaustedGenerator);
+	}
```

Comment:
The `and` prefix is unique to the above four methods which I think could look odd in a sequence:

```java
RetryBuilder retryBuilder = Retry
		.max(1)
		.andDoBeforeRetry(rs -> ...)
		.andDoBeforeRetry(rs -> ...)
		.transientErrors(true);
```

It seems to ready okay (better?) without the prefix:
```java
RetryBuilder retryBuilder = Retry
		.max(1)
		.doBeforeRetry(rs -> ...)
		.doAfterRetry(rs -> ...)
		.transientErrors(true);
```

Also I wonder if `andDelayRetryWith` and `andRetryThen` names could be more aligned with the sync hooks, or with each other at least. Maybe `doBeforeWithDelay` and `doAfterWithDelay` or `doBeforeRetryAsync` and `doAfterRetryAsync`?

Reply:
I like `do{Before|After}Retry[Async]`